kernel/task/src/lib.rs

```diff
@@ -1181,88 +1159,152 @@ impl Drop for JoinableTaskRef {
-    #[doc(hidden)]
     pub fn mark_as_killed(&self, reason: KillReason) -> Result<(), &'static str> {
-        if with_current_task(|t| t == self)
-            .unwrap_or(false)
-        {
-            self.internal_exit(ExitValue::Killed(reason))
-        } else {
-            Err("`mark_as_exited()` can only be invoked on the current task, not on another task.")
+        self.internal_exit(ExitValue::Killed(reason))
+    }
+
+    /// Reaps this task (if orphaned) by taking and dropping its exit value and removing it
+    /// from the system task list.
+    ///
+    /// If this task has *not* been orphaned, meaning it is still joinable,
+    /// then this function does nothing.
+    pub fn reap_if_orphaned(&self) {
+        if !self.is_joinable() {
+            // trace!("Reaping orphaned task... {:?}", self);
+            let _exit_value = self.task.reap_exit_value();
+            // trace!("Reaped orphaned task {:?}, {:?}", self, _exit_value);
         }
     }
 
+    /// Perform any actions needed after a context switch.
+    /// 
+    /// Currently this only does two things:
+    /// 1. Drops any data that the original previous task (before the context switch)
+    ///    prepared for us to drop, as specified by `TaskInner::drop_after_task_switch`.
+    /// 2. Obtains the preemption guard such that preemption can be re-enabled
+    ///    when it is appropriate to do so.
+    ///
+    /// Note: this publicly re-exports the private `TaskRef::post_context_switch_action()`
+    ///       function for use in the early `spawn::task_wrapper` functions,
+    ///       which is the only place where an `ExitableTaskRef` can be obtained. 
+    pub fn post_context_switch_action(&self) -> PreemptionGuard {
+        self.task.post_context_switch_action()
+    }
+
+    /// Allows the unwinder to obtain an `ExitableTaskRef` in order for it to
+    /// be able to invoke this task's [`FailureCleanupFunction`].
+    #[doc(hidden)]
+    pub fn obtain_for_unwinder(current_task: TaskRef) -> Self {
+        Self { task: current_task }
+    }
+}
+
+/// A shareable, cloneable reference to a `Task` that exposes more methods
+/// for task management and auto-derefs into an immutable `&Task` reference.
+/// 
+/// The `TaskRef` type is necessary because in many places across Theseus,
+/// a reference to a Task is used. 
+/// For example, task lists, task spawning, task management, scheduling, etc. 
+/// 
+/// ## Equality comparisons
+/// `TaskRef` implements the [`PartialEq`] and [`Eq`] traits to ensure that
+/// two `TaskRef`s are considered equal if they point to the same underlying `Task`.
+#[derive(Clone)]
+pub struct TaskRef(Arc<TaskRefInner>);
+impl fmt::Debug for TaskRef {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TaskRef")
+            .field("task", &self.0.task)
+            .finish_non_exhaustive()
+    }
+}
+struct TaskRefInner {
+    task: Task,
+    exit_value_mailbox: Mutex<Option<ExitValue>>,
```

Comment:
Why is moving the exit value out `Task` beneficial?

Reply:
1. I eventually want to make it statically typed (with generic type params), but I don't want to have to parameterize the `Task` struct
2. Design-wise, I dislike huge monolithic task structs
3. The smaller the `Task` struct, the cheaper it becomes to allocate and keep a million of em in memory

I realize this doesn't truly move it out of `Task` since it's still within `TaskRef`, but it's a partial step for now.

Reply:
1. Does that mean all task refs would have to have generics? I don't see how that's any better than `Task` having generics, as pretty much all crates use `TaskRef`s rather than `Task`s. A compromise would be only to have `JoinableTaskRef` parameterised like in `dreadnought`.
2. As I said above, `TaskRef`s are pretty much synonymous with `Task`s.
3.
```rust
pub static TASKLIST: MutexIrqSafe<BTreeMap<usize, TaskRef>> = MutexIrqSafe::new(BTreeMap::new());
```
We allocate a `TaskRef`, not `Task`, so moving fields into `TaskRef` doesn't change anything regarding allocation cost.

It just feels weird for some fields to be stored in the `Task` and others to be stored in the `TaskRef`. The only difference with `exit_value` is that `spawn` does not modify it during initialisation e.g. `joinable` could also be moved into the `TaskRef`.

Reply:
As you said, we could parameterize only `Task` and not `TaskRef`, but I'm not 100% settled on a specific design. If you notice in the current main branch, there is a commented-out type called `TaskJoiner` which was my original idea to parameterize a task/join-related type with those generic types. The key design point is that (at least currently) only the `spawn`-related functions know the generic types `F`, `A`, and `R` related to a task's entry point function, and I don't currently see a reason to push that knowledge into `task`.

>  It just feels weird for some fields to be stored in the Task and others to be stored in the TaskRef. The only difference with exit_value is that spawn does not modify it during initialisation e.g. joinable could also be moved into the TaskRef.

Agreed, `joinable` will be removed entirely soon since there are other ways to realize that feature.

Reply:
Another guiding design principle here is the principle of least privilege. Essentially, I don't want any old `TaskRef` to have the privilege (ability) to access or modify an exit value; the easiest way to achieve this is to move it out of the `TaskRef` struct.
As I said before, while this PR alone doesn't achieve that, it's a stepping stone. Eventually it will be entirely removed from `TaskRef` too.